src/Database/Table/ColumnAccessCache.php

```diff
@@ -0,0 +1,200 @@
+<?php
+
+/**
+ * This file is part of the Nette Framework (https://nette.org)
+ * Copyright (c) 2004 David Grudl (https://davidgrudl.com)
+ */
+
+declare(strict_types=1);
+
+namespace Nette\Database\Table;
+
+use Nette\Caching\Cache;
+use Nette\Caching\IStorage;
+
+
+class ColumnAccessCache
+{
+	/** @var Selection */
+	protected $selection;
+
+	/** @var Cache */
+	protected $cache;
+
+	/** @var string */
+	protected $generalCacheKey;
+
+	/** @var string */
+	protected $specificCacheKey;
+
+	/** @var array of touched columns */
+	protected $accessedColumns = [];
+
+	/** @var array of earlier touched columns */
+	protected $previousAccessedColumns;
+
+	/** @var Selection instance of observed accessed columns */
+	protected $observeCache;
+
+
+	public function __construct(Selection $selection, IStorage $cacheStorage = null)
+	{
+		$this->selection = $selection;
+		$this->cache = $cacheStorage ? new Cache($cacheStorage, 'Nette.Database.' . md5($selection->getConnection()->getDsn())) : null;
+	}
+
+
+	public function getStorage(): ?IStorage
+	{
+		return $this->cache ? $this->cache->getStorage() : null;
+	}
+
+
+	/**
+	 * Returns general cache key independent on query parameters or sql limit
+	 * Used e.g. for previously accessed columns caching
+	 */
+	public function getGeneralCacheKey(): string
+	{
+		if ($this->generalCacheKey) {
+			return $this->generalCacheKey;
+		}
+
+		$key = [__CLASS__, $this->selection->getName(), $this->selection->getSqlBuilder()->getConditions()];
+		$trace = [];
+		foreach (debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS) as $item) {
+			$trace[] = isset($item['file'], $item['line']) ? $item['file'] . $item['line'] : null;
+		}
+
+		$key[] = $trace;
+		return $this->generalCacheKey = md5(serialize($key));
+	}
+
+
+	public function setGeneralCacheKey(?string $key): void
+	{
+		$this->generalCacheKey = $key;
+	}
+
+
+	/**
+	 * Returns object specific cache key dependent on query parameters
+	 * Used e.g. for reference memory caching
+	 */
+	public function getSpecificCacheKey(): string
+	{
+		if ($this->specificCacheKey) {
+			return $this->specificCacheKey;
+		}
+
+		return $this->specificCacheKey = $this->selection->getSqlBuilder()->getSelectQueryHash($this->getPreviousAccessedColumns());
+	}
+
+
+	public function setSpecificCacheKey(?string $key): void
+	{
+		$this->specificCacheKey = $key;
+	}
+
+
+	public function getAccessedColumns(): array
+	{
+		return $this->accessedColumns;
+	}
+
+
+	public function setAccessedColumns(array $accessedColumns): void
+	{
+		if (!$this->cache) {
+			return;
+		}
+
+		$this->accessedColumns = $accessedColumns;
+	}
+
+
+	public function setAccessedColumn(string $key, bool $value): void
+	{
+		if (!$this->cache) {
```

Comment:
Just such a stupid thing, maybe positive condition can be used here and in `setAccessedColumns`

````php
if ($this->cache) {
     $this->accessedColumns[$key] = $value;
}
```

Reply:
Fixed on both places.